### PR TITLE
allow decorate model init with saving hparams

### LINKED ISSUE #4525: save_hyperparameters() doesn't work with decorated __init__()
## 🐛 Bug

save_hyperparameters() breaks when __init__() is decorated

https://colab.research.google.com/drive/1njXP32G3FSg4nWvwo7kuhVr6iqH3eabQ?usp=sharing

### Expected behavior

hyperparameters should be saved

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.0.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020
```

### Additional context

I think the right way to detect parameters is using `inspect.signature(cls.__init__).parameters` instead of `inspect.getfullargspec(Reinforce.__init__)`. `signature()` does cooperate with `functools.wraps()`. You can use `kind` attribute on Parameter object to detect the variable positional & keyword arguments.

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Allow decorating Model init and still be able to save hparams

Resolves #4525

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/test_hparams.py [hunk 1]
```diff
@@ -29,20 +30,44 @@
...
+    """ Tests that a model can take an object """
+    @decorate
+    def __init__(self, hparams, *my_args, **my_kwargs):
+        super().__init__()
+        self.save_hyperparameters(hparams)
+
+
+class AssignHparamsDecoratedModel(BoringModel):
+    """ Tests that a model can take an object with explicit setter"""
+    @decorate
```
Comment (tchaton) [inline tests/models/test_hparams.py:65]: Can you try with multiple decorator ? Would undecorated help there  ?

Comment (tchaton) [reply tests/models/test_hparams.py:65]: https://github.com/PyTorchLightning/pytorch-lightning/blob/52a82394d531ad5f6f71327a8194632b06935e1a/tests/backends/launcher.py#L76

**Review by tchaton — CHANGES_REQUESTED** [review]
Great addition ! Can you just test out several layer decorator ?

File: tests/models/test_hparams.py (hunk 1 above)
Comment (Borda) [reply tests/models/test_hparams.py:66]: added double decoration and works...
it may work but it does not simplify the getting keys words :]

**Review by tchaton — APPROVED** [review]
Awesome PR !

File: pytorch_lightning/utilities/parsing.py [hunk 2]
```diff
@@ -79,23 +79,46 @@ def clean_namespace(hparams):
...
+    init_parameters = inspect.signature(cls.__init__).parameters
+    # docs claims the params are always ordered
+    # https://docs.python.org/3/library/inspect.html#inspect.Signature.parameters
+    init_params = list(init_parameters.values())
+    # self is always first
+    n_self = init_params[0].name
+
+    def _get_first_if_any(params, str_type):
+        all_vars = [p for p in params if str(p.kind) == str_type]
+        return all_vars[0].name if all_vars else None
```
Comment (carmocca) [inline pytorch_lightning/utilities/parsing.py:100]: This is very minor so feel free to ignore it but, if you are only interested in the first value, why not just use a loop?

```suggestion
        for p in params:
            if str(p.kind) == str_type:
                return p.name
```

File: pytorch_lightning/utilities/parsing.py [hunk 3]
```diff
@@ -79,23 +79,46 @@ def clean_namespace(hparams):
...
+
+    n_args = _get_first_if_any(init_params, 'VAR_POSITIONAL')
+    n_kwargs = _get_first_if_any(init_params, 'VAR_KEYWORD')
```
Comment (carmocca) [inline pytorch_lightning/utilities/parsing.py:103]: Why not use the enums? You wouldn't have to cast to str above

```suggestion
    n_args = _get_first_if_any(init_params, inspect.Parameter.VAR_POSITIONAL)
    n_kwargs = _get_first_if_any(init_params, inspect.Parameter.VAR_KEYWORD)
```

File: pytorch_lightning/utilities/parsing.py [hunk 4]
```diff
@@ -79,23 +79,46 @@ def clean_namespace(hparams):
...
-    )
+    self_var, args_var, kwargs_var = parse_class_init_keys(cls)
+    filtered_vars = [n for n in (self_var, args_var, kwargs_var) if n]
+    exclude_argnames = (*filtered_vars, '__class__', 'frame', 'frame_args')
 
     # only collect variables that appear in the signature
     local_args = {k: local_vars[k] for k in init_parameters.keys()}
-    local_args.update(local_args.get(kwargs_identifier, {}))
+    if kwargs_var:
+        local_args.update(local_args.get(kwargs_var, {}))
```
Comment (carmocca) [inline pytorch_lightning/utilities/parsing.py:121]: Do you need the `if`? `get` can deal with `None`

```suggestion
    local_args.update(local_args.get(kwargs_var, {}))
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -48,8 +48,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
-- Fixed `setup` callback hook to correctly pass the LightningModule through ([#4608](https://github.com/PyTorchLightning/pytorch-lightning/pull/4608))
+- Allowing decorate model init with saving `hparams` inside ([#4662](https://github.com/PyTorchLightning/pytorch-lightning/pull/4662))
```
Comment (Borda) [inline CHANGELOG.md:52]:
```suggestion


```

File: pytorch_lightning/utilities/parsing.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/utilities/parsing.py]: ok turned to be a bug here...

**Review by SeanNaren — APPROVED** [review]
Curious on the use case, but LGTM!

**Review by SkafteNicki — APPROVED** [review]
LGTM
